packages/react-reconciler/src/ReactFiber.js

```diff
@@ -278,8 +281,31 @@ const createFiber = function(
   return new FiberNode(tag, pendingProps, key, mode);
 };
 
-function shouldConstruct(Component) {
-  return !!(Component.prototype && Component.prototype.isReactComponent);
+function shouldConstruct(Component: Function) {
+  const prototype = Component.prototype;
+  return (
+    typeof prototype === 'object' &&
+    prototype !== null &&
+    typeof prototype.isReactComponent === 'object' &&
+    prototype.isReactComponent !== null
+  );
+}
+
+export function reassignLazyComponentTag(
```

Comment:
Slight preference for returning the tag and assigning it in the outer function to avoid an unnecessary Get since this will be inlined.